shuup/discounts/utils.py

```diff
@@ -57,17 +64,98 @@ def get_potential_discounts_for_product(context, product):
     else:
         condition_query &= Q(coupon_code__isnull=True)
 
+    from shuup.discounts.models import Discount
+    if available_only:
+        base_queryset = Discount.objects.available(shop)
+    else:
+        base_queryset = Discount.objects.filter(shops=shop, active=True)
+
     # Get all possible discounts for the current product and context
-    return Discount.objects.available(shop).filter(
-        condition_query
-    ).values_list(
-        "discounted_price_value", "discount_amount_value", "discount_percentage", "coupon_code__code"
-    ).distinct()
+    return base_queryset.filter(condition_query).distinct()
 
 
 def get_active_discount_for_code(order_or_order_source, code):
+    from shuup.discounts.models import Discount
     shop = order_or_order_source.shop
     return Discount.objects.available(shop).filter(
         Q(coupon_code__active=True) &
         Q(coupon_code__code__iexact=code)
     ).first()
+
+
+def get_next_dates_for_range(weekday, from_hour, to_hour):
+    """
+    Get datetime ranges for the next weekday
+
+    Example:
+        Give me the date ranges for the next Sunday from 1pm to 10pm
+        It will return a tuple of datetimes.
+
+    If the requested weekday is the same of today, it will return both the ranges for today
+        and also for the next week.
+
+    :rtype list[datetime.datetime]
+    """
+    import datetime
+    from django.utils.timezone import now
+
+    now_datetime = now()
+    next_date = now_datetime + datetime.timedelta(days=(abs(weekday - now_datetime.weekday()) % 7))
+    ranges = [
+        next_date.replace(hour=from_hour.hour, minute=from_hour.minute),
+        next_date.replace(hour=to_hour.hour, minute=to_hour.minute)
+    ]
+
+    # the next date is the same as today, let's return also the next week ranges
+    if next_date.date() == now().date():
+        next_week_date = next_date + datetime.timedelta(days=7)
+        ranges.extend([
+            next_week_date.replace(hour=from_hour.hour, minute=from_hour.minute),
+            next_week_date.replace(hour=to_hour.hour, minute=to_hour.minute)
+        ])
+
+    return ranges
+
+
+def get_price_expiration(context, product):
```

Comment:
Problem! This doesn't look like being lighting fast operation. So why not return shared expiration date for the whole module based on active discounts. Now every time price for product expires we need to this "expensive" expire on calculation. This "shared" expiration date then is easy to cache and can be used for all prices coming through this discount module.

Likely the shop discounts are not timed the way that those would change in every 5min anyways.